static/app/actionCreators/pageFilters.tsx

```diff
@@ -322,58 +332,61 @@ export function updateParams(obj: UrlParams, router?: Router, options?: Options)
 }
 
 /**
- * Creates a new query parameter object given new params and old params
- * Preserves the old query params, except for `cursor` (can be overriden with keepCursor option)
+ * Merges an UpdateParams object into a Location['query'] object. Results in a
+ * PageFilterQuery
+ *
+ * Preserves the old query params, except for `cursor` (can be overriden with
+ * keepCursor option)
  *
  * @param obj New query params
- * @param oldQueryParams Old query params
+ * @param currentQuery The current query parameters
  * @param [options] Options object
  */
 function getNewQueryParams(
-  obj: UrlParams,
-  oldQueryParams: UrlParams,
-  {resetParams, keepCursor}: Options = {}
+  obj: UpdateParams,
+  currentQuery: Location['query'],
+  options: Options = {}
 ) {
-  const {cursor, statsPeriod, ...oldQuery} = oldQueryParams;
-  const oldQueryWithoutResetParams = !!resetParams?.length
-    ? omit(oldQuery, resetParams)
-    : oldQuery;
-
-  const newQuery = getParams({
-    ...oldQueryWithoutResetParams,
-    // Some views update using `period`, and some `statsPeriod`, we should make this uniform
-    period: !obj.start && !obj.end ? obj.period || statsPeriod : null,
-    ...obj,
-  });
+  const {resetParams, keepCursor} = options;
 
-  if (newQuery.start) {
-    newQuery.start = getUtcDateString(newQuery.start);
-  }
+  const cleanCurrentQuery = !!resetParams?.length
+    ? omit(currentQuery, resetParams)
+    : currentQuery;
 
-  if (newQuery.end) {
-    newQuery.end = getUtcDateString(newQuery.end);
-  }
+  // Normalize existing query parameters
+  const currentQueryState = getStateFromQuery(cleanCurrentQuery, {
+    allowEmptyPeriod: true,
+  });
 
-  if (keepCursor) {
-    newQuery.cursor = cursor;
-  }
+  // Extract non page filter parameters.
+  const cursorParam = !keepCursor ? 'cursor' : null;
+  const omittedParameters = [...Object.values(URL_PARAM), cursorParam].filter(defined);
+
+  const extraParams = omit(cleanCurrentQuery, omittedParameters);
+
+  // Combine parameters
+  const project = obj.project || currentQueryState.project;
+  const environment = obj.environment || currentQueryState.environment;
+  const start = obj.start || currentQueryState.start;
+  const end = obj.end || currentQueryState.end;
+  const utc = obj.utc || currentQueryState.utc;
+
+  // `period` is deprecated as a url parameter, though some pages may still
+  // include it (need to actually validate this?), normalize period and stats
+  // period together
+  const statsPeriod = obj.statsPeriod || obj.period || currentQueryState.period;
+
+  const newQuery: PageFilterQuery = {
+    project: project?.map(String),
+    environment,
+    start: statsPeriod ? null : start instanceof Date ? getUtcDateString(start) : start,
+    end: statsPeriod ? null : end instanceof Date ? getUtcDateString(end) : end,
```

Comment:
@evanpurkhiser We should probably check if the dates here are valid, the instanceof new Date("invalid date") will still pass here otherwise. UX wise, I'm not sure if it's best to discard invalid dates or throw an error. I'm in favour of throwing so that in case this happens we have visibility and can fix it.

Reply:
We probably should check somewhere else if the dates are valid or not.

This code is a bit of a rats nest, before it wasn't even handling string-ification of date objects afaict.

For now I am OK with leaving this as it is, since it's at least a big improvement over how it worked before 😬

Reply:
Makes sense - I'm lacking context on some of these components and how they all work together in the greater deal of things, so I may say something dumb every now and then 😅 Def a nice improvement from before though 🙏🏼